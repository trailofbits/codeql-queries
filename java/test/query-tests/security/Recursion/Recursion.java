import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.Map;

class StreamException extends RuntimeException {
    public StreamException(Throwable cause) {
        super(cause);
    }
}

class Token {
    public static final int TYPE_MAP_ID_TO_VALUE = 0x2;
    
    private int type;
    private String id;
    private Object value;
    
    public int getType() {
        return type;
    }
    
    public String getId() {
        return id;
    }
    
    public Object getValue() {
        return value;
    }
}

class TokenFormatter {
    public Token read(InputStreamReader in) throws IOException {
        // Implementation would go here
        return new Token();
    }
}

class RecursiveCallExample {
    private TokenFormatter tokenFormatter = new TokenFormatter();
    private Map<String, Object> idRegistry = new HashMap<>();
    private InputStreamReader in;

    public RecursiveCallExample(InputStreamReader in) {
        this.in = in;
    }

    // finding: readToken calls itself
    // Based on https://github.com/x-stream/xstream/blob/dfa1d35462fe84412ee72a9b0cf5b5c633086520/xstream/src/java/com/thoughtworks/xstream/io/binary/BinaryStreamReader.java#L165
    Token readToken() {
        // ...
        try {
            final Token token = tokenFormatter.read(in);
            switch (token.getType()) {
            case Token.TYPE_MAP_ID_TO_VALUE: // 0x2
                idRegistry.put(token.getId(), token.getValue());
                return readToken(); // Next one please.
            default:
                return token;
            }
        } catch (final IOException e) {
            throw new StreamException(e);
        }
        // ...
    }
}

class RecursiveCallBasic {
    // finding: foo-bar recursive loop
    public boolean bar() {
        boolean fooResult = foo();
        return fooResult;
    }

    public boolean foo() {
        return bar();
    }

    // finding: calls to self
    public boolean directRecursive() {
        return directRecursive();
    }

    // finding: level0->level1->level2->level0
    public boolean level0() {
        if (someCondition()) {
            return true;
        }        
        return level1();
    }
    public boolean level1() {
        if (someCondition()) {
            return true;
        }        
        return level2();
    }
    public boolean level2() {
        if (someCondition()) {
            return true;
        }
        return level0();
    }

    private boolean someCondition() {
        return false;
    }
}

class RecursiveCallNonLinear {
    // finding: level0->...->level0
    public boolean level0() {
        if (someOtherCondition()) {
            return true;
        }
        if (someCondition()) {
            return level1();
        }        
        return level2();
    }
    public boolean level1() {
        if (someCondition()) {
            return true;
        }        
        return level2();
    }
    public boolean level2() {
        if (someCondition()) {
            return level1();
        }
        return level0();
    }

    private boolean someCondition() {
        return false;
    }

    private boolean someOtherCondition() {
        return true;
    }
}

class RecursiveCallWronglyLimited {
    // finding: recursion is not limited
    public boolean directRecursiveNoDepth(int anything, int depth) {
        if (depth == 0) {
            return true;
        }
        return directRecursiveNoDepth(anything, depth);
    }
}

class RecursiveCallLimited {
    // todook: recursion is limited
    public boolean directRecursiveDepth(int depth) {
        if (depth == 0) {
            return true;
        }
        return directRecursiveDepth(depth - 1);
    }

    // todook: level0->level1->level2->level0 with bound
    public boolean level0D(int depth) {
        if (depth == 0) {
            return true;
        }
        if (someCondition()) {
            return true;
        }        
        return level1D(depth-1);
    }
    public boolean level1D(int depth) {
        if (someCondition()) {
            return true;
        }        
        return level2D(depth);
    }
    public boolean level2D(int depth) {
        if (someCondition()) {
            return true;
        }
        return level0D(depth-1);
    }

    private boolean someCondition() {
        return false;
    }
}
 
// ok
class NotRecursive {
    public static boolean foo() {
        return bar();
    }

    public static boolean bar() {
        return true;
    }
}