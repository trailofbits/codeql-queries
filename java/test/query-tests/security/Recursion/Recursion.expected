edges
| Recursion.java:57:24:57:34 | readToken calls readToken : Token | Recursion.java:57:24:57:34 | readToken calls readToken | provenance |  |
| Recursion.java:57:24:57:34 | readToken calls readToken : Token | Recursion.java:57:24:57:34 | readToken calls readToken : Token | provenance |  |
| Recursion.java:81:16:81:32 | directRecursive calls directRecursive : Boolean | Recursion.java:81:16:81:32 | directRecursive calls directRecursive | provenance |  |
| Recursion.java:81:16:81:32 | directRecursive calls directRecursive : Boolean | Recursion.java:81:16:81:32 | directRecursive calls directRecursive : Boolean | provenance |  |
| Recursion.java:89:16:89:23 | level0 calls level1 : Boolean | Recursion.java:101:16:101:23 | level2 calls level0 : Boolean | provenance |  |
| Recursion.java:95:16:95:23 | level1 calls level2 : Boolean | Recursion.java:89:16:89:23 | level0 calls level1 | provenance |  |
| Recursion.java:95:16:95:23 | level1 calls level2 : Boolean | Recursion.java:89:16:89:23 | level0 calls level1 : Boolean | provenance |  |
| Recursion.java:101:16:101:23 | level2 calls level0 : Boolean | Recursion.java:95:16:95:23 | level1 calls level2 : Boolean | provenance |  |
| Recursion.java:116:20:116:27 | level0 calls level1 : Boolean | Recursion.java:130:16:130:23 | level2 calls level0 | provenance |  |
| Recursion.java:116:20:116:27 | level0 calls level1 : Boolean | Recursion.java:130:16:130:23 | level2 calls level0 : Boolean | provenance |  |
| Recursion.java:118:16:118:23 | level0 calls level2 : Boolean | Recursion.java:130:16:130:23 | level2 calls level0 | provenance |  |
| Recursion.java:118:16:118:23 | level0 calls level2 : Boolean | Recursion.java:130:16:130:23 | level2 calls level0 : Boolean | provenance |  |
| Recursion.java:124:16:124:23 | level1 calls level2 : Boolean | Recursion.java:116:20:116:27 | level0 calls level1 | provenance |  |
| Recursion.java:124:16:124:23 | level1 calls level2 : Boolean | Recursion.java:116:20:116:27 | level0 calls level1 : Boolean | provenance |  |
| Recursion.java:124:16:124:23 | level1 calls level2 : Boolean | Recursion.java:128:20:128:27 | level2 calls level1 | provenance |  |
| Recursion.java:124:16:124:23 | level1 calls level2 : Boolean | Recursion.java:128:20:128:27 | level2 calls level1 : Boolean | provenance |  |
| Recursion.java:128:20:128:27 | level2 calls level1 : Boolean | Recursion.java:118:16:118:23 | level0 calls level2 : Boolean | provenance |  |
| Recursion.java:128:20:128:27 | level2 calls level1 : Boolean | Recursion.java:124:16:124:23 | level1 calls level2 : Boolean | provenance |  |
| Recursion.java:130:16:130:23 | level2 calls level0 : Boolean | Recursion.java:118:16:118:23 | level0 calls level2 : Boolean | provenance |  |
| Recursion.java:130:16:130:23 | level2 calls level0 : Boolean | Recursion.java:124:16:124:23 | level1 calls level2 : Boolean | provenance |  |
| Recursion.java:148:16:148:54 | directRecursiveNoDepth calls directRecursiveNoDepth : Boolean | Recursion.java:148:16:148:54 | directRecursiveNoDepth calls directRecursiveNoDepth | provenance |  |
| Recursion.java:148:16:148:54 | directRecursiveNoDepth calls directRecursiveNoDepth : Boolean | Recursion.java:148:16:148:54 | directRecursiveNoDepth calls directRecursiveNoDepth : Boolean | provenance |  |
| Recursion.java:177:16:177:47 | directRecLocalFlow calls directRecLocalFlow : Boolean | Recursion.java:177:16:177:47 | directRecLocalFlow calls directRecLocalFlow | provenance |  |
| Recursion.java:177:16:177:47 | directRecLocalFlow calls directRecLocalFlow : Boolean | Recursion.java:177:16:177:47 | directRecLocalFlow calls directRecLocalFlow : Boolean | provenance |  |
nodes
| Recursion.java:57:24:57:34 | readToken calls readToken | semmle.label | readToken calls readToken |
| Recursion.java:57:24:57:34 | readToken calls readToken : Token | semmle.label | readToken calls readToken : Token |
| Recursion.java:71:29:71:33 | bar calls foo | semmle.label | bar calls foo |
| Recursion.java:76:16:76:20 | foo calls bar | semmle.label | foo calls bar |
| Recursion.java:81:16:81:32 | directRecursive calls directRecursive | semmle.label | directRecursive calls directRecursive |
| Recursion.java:81:16:81:32 | directRecursive calls directRecursive : Boolean | semmle.label | directRecursive calls directRecursive : Boolean |
| Recursion.java:89:16:89:23 | level0 calls level1 | semmle.label | level0 calls level1 |
| Recursion.java:89:16:89:23 | level0 calls level1 : Boolean | semmle.label | level0 calls level1 : Boolean |
| Recursion.java:95:16:95:23 | level1 calls level2 : Boolean | semmle.label | level1 calls level2 : Boolean |
| Recursion.java:101:16:101:23 | level2 calls level0 : Boolean | semmle.label | level2 calls level0 : Boolean |
| Recursion.java:116:20:116:27 | level0 calls level1 | semmle.label | level0 calls level1 |
| Recursion.java:116:20:116:27 | level0 calls level1 : Boolean | semmle.label | level0 calls level1 : Boolean |
| Recursion.java:118:16:118:23 | level0 calls level2 | semmle.label | level0 calls level2 |
| Recursion.java:118:16:118:23 | level0 calls level2 : Boolean | semmle.label | level0 calls level2 : Boolean |
| Recursion.java:124:16:124:23 | level1 calls level2 | semmle.label | level1 calls level2 |
| Recursion.java:124:16:124:23 | level1 calls level2 : Boolean | semmle.label | level1 calls level2 : Boolean |
| Recursion.java:128:20:128:27 | level2 calls level1 | semmle.label | level2 calls level1 |
| Recursion.java:128:20:128:27 | level2 calls level1 : Boolean | semmle.label | level2 calls level1 : Boolean |
| Recursion.java:130:16:130:23 | level2 calls level0 | semmle.label | level2 calls level0 |
| Recursion.java:130:16:130:23 | level2 calls level0 : Boolean | semmle.label | level2 calls level0 : Boolean |
| Recursion.java:148:16:148:54 | directRecursiveNoDepth calls directRecursiveNoDepth | semmle.label | directRecursiveNoDepth calls directRecursiveNoDepth |
| Recursion.java:148:16:148:54 | directRecursiveNoDepth calls directRecursiveNoDepth : Boolean | semmle.label | directRecursiveNoDepth calls directRecursiveNoDepth : Boolean |
| Recursion.java:177:16:177:47 | directRecLocalFlow calls directRecLocalFlow | semmle.label | directRecLocalFlow calls directRecLocalFlow |
| Recursion.java:177:16:177:47 | directRecLocalFlow calls directRecLocalFlow : Boolean | semmle.label | directRecLocalFlow calls directRecLocalFlow : Boolean |
subpaths
#select
| Recursion.java:57:24:57:34 | readToken calls readToken | Recursion.java:57:24:57:34 | readToken calls readToken | Recursion.java:57:24:57:34 | readToken calls readToken | Found a recursion:  |
| Recursion.java:57:24:57:34 | readToken calls readToken | Recursion.java:57:24:57:34 | readToken calls readToken : Token | Recursion.java:57:24:57:34 | readToken calls readToken | Found a recursion:  |
| Recursion.java:71:29:71:33 | bar calls foo | Recursion.java:71:29:71:33 | bar calls foo | Recursion.java:71:29:71:33 | bar calls foo | Found a recursion:  |
| Recursion.java:76:16:76:20 | foo calls bar | Recursion.java:76:16:76:20 | foo calls bar | Recursion.java:76:16:76:20 | foo calls bar | Found a recursion:  |
| Recursion.java:81:16:81:32 | directRecursive calls directRecursive | Recursion.java:81:16:81:32 | directRecursive calls directRecursive | Recursion.java:81:16:81:32 | directRecursive calls directRecursive | Found a recursion:  |
| Recursion.java:81:16:81:32 | directRecursive calls directRecursive | Recursion.java:81:16:81:32 | directRecursive calls directRecursive : Boolean | Recursion.java:81:16:81:32 | directRecursive calls directRecursive | Found a recursion:  |
| Recursion.java:89:16:89:23 | level0 calls level1 | Recursion.java:101:16:101:23 | level2 calls level0 : Boolean | Recursion.java:89:16:89:23 | level0 calls level1 | Found a recursion:  |
| Recursion.java:116:20:116:27 | level0 calls level1 | Recursion.java:128:20:128:27 | level2 calls level1 : Boolean | Recursion.java:116:20:116:27 | level0 calls level1 | Found a recursion:  |
| Recursion.java:116:20:116:27 | level0 calls level1 | Recursion.java:130:16:130:23 | level2 calls level0 : Boolean | Recursion.java:116:20:116:27 | level0 calls level1 | Found a recursion:  |
| Recursion.java:118:16:118:23 | level0 calls level2 | Recursion.java:118:16:118:23 | level0 calls level2 | Recursion.java:118:16:118:23 | level0 calls level2 | Found a recursion:  |
| Recursion.java:124:16:124:23 | level1 calls level2 | Recursion.java:124:16:124:23 | level1 calls level2 | Recursion.java:124:16:124:23 | level1 calls level2 | Found a recursion:  |
| Recursion.java:128:20:128:27 | level2 calls level1 | Recursion.java:128:20:128:27 | level2 calls level1 | Recursion.java:128:20:128:27 | level2 calls level1 | Found a recursion:  |
| Recursion.java:128:20:128:27 | level2 calls level1 | Recursion.java:128:20:128:27 | level2 calls level1 : Boolean | Recursion.java:128:20:128:27 | level2 calls level1 | Found a recursion:  |
| Recursion.java:128:20:128:27 | level2 calls level1 | Recursion.java:130:16:130:23 | level2 calls level0 : Boolean | Recursion.java:128:20:128:27 | level2 calls level1 | Found a recursion:  |
| Recursion.java:130:16:130:23 | level2 calls level0 | Recursion.java:128:20:128:27 | level2 calls level1 : Boolean | Recursion.java:130:16:130:23 | level2 calls level0 | Found a recursion:  |
| Recursion.java:130:16:130:23 | level2 calls level0 | Recursion.java:130:16:130:23 | level2 calls level0 | Recursion.java:130:16:130:23 | level2 calls level0 | Found a recursion:  |
| Recursion.java:130:16:130:23 | level2 calls level0 | Recursion.java:130:16:130:23 | level2 calls level0 : Boolean | Recursion.java:130:16:130:23 | level2 calls level0 | Found a recursion:  |
| Recursion.java:148:16:148:54 | directRecursiveNoDepth calls directRecursiveNoDepth | Recursion.java:148:16:148:54 | directRecursiveNoDepth calls directRecursiveNoDepth | Recursion.java:148:16:148:54 | directRecursiveNoDepth calls directRecursiveNoDepth | Found a recursion:  |
| Recursion.java:148:16:148:54 | directRecursiveNoDepth calls directRecursiveNoDepth | Recursion.java:148:16:148:54 | directRecursiveNoDepth calls directRecursiveNoDepth : Boolean | Recursion.java:148:16:148:54 | directRecursiveNoDepth calls directRecursiveNoDepth | Found a recursion:  |
| Recursion.java:177:16:177:47 | directRecLocalFlow calls directRecLocalFlow | Recursion.java:177:16:177:47 | directRecLocalFlow calls directRecLocalFlow | Recursion.java:177:16:177:47 | directRecLocalFlow calls directRecLocalFlow | Found a recursion:  |
| Recursion.java:177:16:177:47 | directRecLocalFlow calls directRecLocalFlow | Recursion.java:177:16:177:47 | directRecLocalFlow calls directRecLocalFlow : Boolean | Recursion.java:177:16:177:47 | directRecLocalFlow calls directRecLocalFlow | Found a recursion:  |
